add: autonat (#278)
libp2p/docs: content/concepts/nat/autonat.md

---
title: "AutoNAT"

> description: "AutoNAT lets peers request dial-backs from peers providing the AutoNAT service."

Reviewer: ```suggestion
description: "AutoNAT lets nodes discover if they are behind a NAT."
```

weight: 200
aliases:
    - "/concepts/autonat"
    - "/concepts/nat/autonat"
---

## Background

While the [identify protocol][spec_identify] lets peers inform each other about their observed network

> addresses, not all networks will allow incoming connections on the same port used for dialing out.

Reviewer: ```suggestion
addresses, not all such addresses are dialable from outside the network. Meaning the peer might be in a private network behind a NAT and unreachable.
```

> To solve this problem, libp2p has implemented a protocol called AutoNAT, which allows nodes to determine
> whether or not they are behind a NAT and, if necessary, find a way to improve their connectivity.

Reviewer: ```suggestion
To prevent this problem of advertising undialable addresses, libp2p has implemented a protocol called AutoNAT, which allows nodes to determine
whether or not they are behind a NAT.
Furthermore, this lays the foundation for nodes to find a way to improve their connectivity to peers in public networks.
```

## What is AutoNAT?

AutoNAT allows a node to request other peers to dial its presumed public addresses. If a few of these

> dial attempts are successful, the node can be reasonably certain that it is not behind a NAT. On the other

Reviewer: ```suggestion
dial attempts are successful, the node can be reasonably ascertain that it is not behind a NAT. On the other
```

hand, if a few of these dial attempts fail, it strongly indicates that a NAT is blocking incoming connections.

The AutoNAT protocol uses the protocol ID `/libp2p/autonat/1.0.0` and involves the exchange of `Dial` and
`DialResponse` messages.

To initiate the protocol, a node sends a `Dial` message to another peer containing
a list of multiaddresses. The peer then attempts to dial these addresses using a different IP and peer ID
than it uses for its regular libp2p connection. If at least one of the dials is successful, the peer sends a
`DialResponse` message with the `ResponseStatus`: `SUCCESS` to the requesting node.

If all dials fail, the peer sends a `DialResponse` message with the `ResponseStatus`: `E_DIAL_ERROR`.
The requesting node can use the response from the peer to determine whether or not it is behind a NAT.
> If the response indicates success, the node is likely not behind a NAT and does not need to use a relay
> server to improve its connectivity. If the response indicates an error, the node is likely behind a NAT

> > and may need to use a relay server to communicate with other nodes in the network.

Reviewer: ```suggestion
> and may need to use a relay server to communicate with other nodes in the network.
```
We should link to the DCUtR document here when mentioning the relay server.

{{< alert icon="" context="caution">}}

> To prevent certain types of attacks, implementations of AutoNAT must not dial any multiaddress that

Reviewer: ```suggestion
To prevent [certain types of attacks](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1), libp2p implementations of AutoNAT must not dial any multiaddress that
```

is not based on the IP address of the requesting node AND must not accept dial requests via relayed

> connections.

Reviewer: ```suggestion
connections (because it is not possible to validate a node's IP address that arrives via a relayed connection).
```

This is to prevent amplification attacks, in which an attacker provides many clients with the same

> faked MAPPED-ADDRESS that points to the intended target, causing all traffic to be focused on the

Reviewer: ```suggestion
faked [MAPPED-ADDRESS](https://www.rfc-editor.org/rfc/rfc3489#section-11.2.1) that points to the intended target, causing all traffic to be focused on the
```

---
title: "AutoNAT"
description: "AutoNAT lets nodes discover if they are behind a NAT."
weight: 200
aliases:
    - "/concepts/autonat"
    - "/concepts/nat/autonat"
---

## Background

While the [identify protocol][spec_identify] lets peers inform each other about their observed network
addresses, however, it is possible that some of these addresses are not accessible from outside the network, as the peer may be located in a private network that is behind a [NAT](overview.md) and therefore unreachable.

To prevent this problem of advertising undialable addresses, libp2p has implemented a protocol called AutoNAT, which allows nodes to determine

> whether or not they are behind a NAT.

Reviewer 2: This is more about determining your position and then taking specific action based on that, for example switching to DHT server mode, or getting a reservation with a relay.
AutoNAT current can't test _individual_ addresses (that's what my AutoNAT v2 proposal would add).

Furthermore, this lays the foundation for nodes to find a way to improve their connectivity to peers in public networks.

## What is AutoNAT?

AutoNAT allows a node to request other peers to dial its presumed public addresses. If a few of these
dial attempts are successful, the node can be reasonably ascertain that it is not behind a NAT. On the other

> hand, if a few of these dial attempts fail, it strongly indicates that a NAT is blocking incoming connections.

Reviewer 2: * if all or most of these dial attempts
